src/CoinEntry.h

```diff
@@ -40,8 +41,11 @@ class CoinEntry {
     // It is optional, only UTXO chains need it, default impl. leaves empty result.
     virtual void plan(TWCoinType coin, const Data& dataIn, Data& dataOut) const { return; }
 
-    // Optional method for obtaining hash(es) for signing, needed for external signing. Hashes are linked to the associated pubkey/pubkeyhash.
-    virtual HashPubkeyList preImageHashes(TWCoinType coin, const Data& txInputData) const { return HashPubkeyList(); }
+    // Optional method for obtaining hash(es) for signing, needed for external signing.
+    // It will return a proto object named `PreSigningOutput` which will include hash.
+    // We provide a default `PreSigningOutput` in TransactionCompiler.proto.
+    // For some special coins, such as bitcoin, we will create a custom `PreSigningOutput` object in its proto file.
+    virtual Data preImageHashes(TWCoinType coin, const Data& txInputData) const { return Data(); }
```

Comment:
I propose to have a C++ struct/class corresponding to PreSigningOutput, and use that within the C++ part, not plain Data. Use the corresponding serialized PreSigningOuput only through the C interface.

Reply:
Do you mean we return a C++ class/struct for `preImageHashes and compile`, but if we do that, for different coins, we have different class or struct, like `SigningOutput`, we may need use class template or function template to do it, will it bring a big change for CoinEntry class?

Reply:
The `PreSigningOutput` proto message, defined in `TransactionCompiler.proto`, is chain-agnostic.  The `preImageHashes` C++ methods return a Data byte stream, containing a serialized version of a `PreSigningOutput` proto message.
I suggest to create a `PreSigningOutput` C++ class equivalent to [`PreSigningOutput`](https://github.com/trustwallet/wallet-core/pull/2113/files#diff-d8f74abc8685d3d6c57f7c3b0481b80fd1bc6d915621705b1ca3a11e08046370R7) (only a few fields), and use that class in the C++ code, and convert to a protobuf message only on the C++/C boundary.
Does this make sense?

Reply:
Yes, you are right. hash/key pair can works for all coins. We determined to use proto message `PreSigningOutput `, because we want to make things more clear when we handle different coins. As we expected, custom proto message can provide more details for some coins, such as solana, we have defined a custom proto mesage for it.
```
message PreSigningOutput {
    /// signer list
    repeated bytes signers = 1;
    /// Pre-image data. There is no hashing for Solana presign image
    bytes data = 2;
    /// error code, 0 is ok, other codes will be treated as errors
    int32 errorCode = 3;
    /// error code description
    string error = 4;
}
```
Although key/pair can work for it, proto message makes it more clear.

What do you think about it? @catenocrypt

Reply:
I can see two kinds of typical cases:
- One signature is needed (simpler chains) There is one digest to sign, typically a hash of the pre-transaction.  It is obvious which private key to use (the one associated to the single account).
- Multiple signatures needed.  There are multiple hashes to sign, possibly with different keys (e.g. for input UTXOs).  All combinations are possible: one hash to be signed with multiple keys; multiple hashes to be signed with the same key, or many-to-many.  As there are multiple keys needed, each hash has to be accompanied by an associated data indicating which key to use, typcially pubkey or pubkey hash (but it's semantics is indeed chain-specific).

So I think a list of hash-to-sign & associated-data covers all cases.

I feel custom protos complicates the scheme too much.

Reply:
Revisiting my comment, doing new review